compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -1689,7 +1689,7 @@ public void setForeachTypedBindingPatternType(BLangForeach foreachNode) {
                 BType constraint = getReferredType(((BXMLType) collectionType).constraint);
                 while (constraint.tag == TypeTags.XML) {
                     collectionType = constraint;
-                    constraint = ((BXMLType) collectionType).constraint;
+                    constraint = getReferredType(((BXMLType) collectionType).constraint);
```

Comment:
Shall we refactor the logic in the `[getReferredType](L1309)` function as follows
```java
public static BType getReferredType(BType type) {
        if (type.tag == TypeTags.TYPEREFDESC) {
            return getReferredType(((BTypeReferenceType) type).referredType);
        }
        return type;
    }
```